// @ariakit/react-store/dist/index.js
"use client";
import { useEvent, useLiveRef, useSafeLayoutEffect } from "@ariakit/react-utils";
import { batch, init, subscribe } from "@ariakit/store";
import { hasOwnProperty, identity } from "@ariakit/utils";
import * as React from "react";
import { useSyncExternalStore } from "use-sync-external-store/shim";
//#region src/index.tsx
const noopSubscribe = () => () => {};
function isStoreKeyArray(value) {
	return Array.isArray(value);
}
function isSameValue(value, other) {
	return value === other || value !== value && other !== other;
}
function hasSameStoreKeys(keys, otherKeys) {
	if (keys?.length !== otherKeys.length) return false;
	for (let index = 0; index < keys.length; index += 1) {
		const key = keys[index];
		if (key === void 0) return false;
		if (isSameValue(key, otherKeys[index])) continue;
		return false;
	}
	return true;
}
function useStableStoreKeys(keys) {
	const keysRef = React.useRef(null);
	const currentKeys = keysRef.current;
	if (keys === null) {
		keysRef.current = null;
		return null;
	}
	if (hasSameStoreKeys(currentKeys, keys)) return currentKeys;
	const nextKeys = [...keys];
	keysRef.current = nextKeys;
	return nextKeys;
}
function useStoreState(store, keyOrKeysOrSelector = identity, selector) {
	const subscriptionKeys = useStableStoreKeys(isStoreKeyArray(keyOrKeysOrSelector) ? keyOrKeysOrSelector : typeof keyOrKeysOrSelector === "function" ? null : [keyOrKeysOrSelector]);
	const storeSubscribe = React.useCallback((callback) => {
		if (!store) return noopSubscribe();
		if (subscriptionKeys?.length === 0) return noopSubscribe();
		return subscribe(store, subscriptionKeys, callback);
	}, [store, subscriptionKeys]);
	const getSnapshot = () => {
		const state = store?.getState();
		if (isStoreKeyArray(keyOrKeysOrSelector)) return selector?.(state);
		if (typeof keyOrKeysOrSelector === "function") return keyOrKeysOrSelector(state);
		if (!state) return;
		if (!hasOwnProperty(state, keyOrKeysOrSelector)) return;
		return state[keyOrKeysOrSelector];
	};
	return useSyncExternalStore(storeSubscribe, getSnapshot, getSnapshot);
}
function getStoreStateObjectKeys(object, selectorKeys) {
	const keys = [];
	let hasSelector = false;
	for (const prop in object) {
		const keyOrSelector = object[prop];
		if (keyOrSelector === void 0) continue;
		if (typeof keyOrSelector === "function") {
			hasSelector = true;
			continue;
		}
		if (keys.includes(keyOrSelector)) continue;
		keys.push(keyOrSelector);
	}
	if (!hasSelector) return keys;
	if (!selectorKeys) return null;
	for (const key of selectorKeys) {
		if (keys.includes(key)) continue;
		keys.push(key);
	}
	return keys;
}
function useStoreStateObject(store, objectOrKeys, object) {
	const stateObject = isStoreKeyArray(objectOrKeys) ? object ?? {} : objectOrKeys;
	const selectorKeys = isStoreKeyArray(objectOrKeys) ? objectOrKeys : void 0;
	const objRef = React.useRef({});
	const subscriptionKeys = useStableStoreKeys(getStoreStateObjectKeys(stateObject, selectorKeys));
	const storeSubscribe = React.useCallback((callback) => {
		if (!store) return noopSubscribe();
		if (subscriptionKeys?.length === 0) return noopSubscribe();
		return subscribe(store, subscriptionKeys, callback);
	}, [store, subscriptionKeys]);
	const getSnapshot = () => {
		const state = store?.getState();
		let updated = false;
		const obj = objRef.current;
		for (const prop in stateObject) {
			const keyOrSelector = stateObject[prop];
			if (keyOrSelector === void 0) continue;
			if (typeof keyOrSelector === "function") {
				const value = keyOrSelector(state);
				if (!Object.is(value, obj[prop])) {
					obj[prop] = value;
					updated = true;
				}
			}
			if (typeof keyOrSelector === "function") continue;
			const value = state && hasOwnProperty(state, keyOrSelector) ? state[keyOrSelector] : void 0;
			if (!Object.is(value, obj[prop])) {
				obj[prop] = value;
				updated = true;
			}
		}
		if (updated) objRef.current = { ...obj };
		return objRef.current;
	};
	return useSyncExternalStore(storeSubscribe, getSnapshot, getSnapshot);
}
/**
* Synchronizes the store with the props, including parent store props.
* @param store The store to synchronize.
* @param props The props to synchronize with.
* @param key The key of the value prop.
* @param setKey The key of the setValue prop.
*/
function useStoreProps(store, props, key, setKey) {
	const value = hasOwnProperty(props, key) ? props[key] : void 0;
	const setValue = setKey ? props[setKey] : void 0;
	const hasSetValue = !!setValue;
	const propsRef = useLiveRef({
		value,
		setValue
	});
	useSafeLayoutEffect(() => {
		if (!hasSetValue) return;
		return subscribe(store, [key], (state, prev) => {
			const { value, setValue } = propsRef.current;
			if (!setValue) return;
			if (isSameValue(state[key], prev[key])) return;
			if (isSameValue(state[key], value)) return;
			setValue(state[key]);
		});
	}, [
		store,
		key,
		hasSetValue
	]);
	useSafeLayoutEffect(() => {
		if (value === void 0) return;
		store.setState(key, value);
		return batch(store, [key], () => {
			if (value === void 0) return;
			store.setState(key, value);
		});
	});
}
/**
* Creates a React store from a core store object and returns a tuple with the
* store and a function to update the store.
* @param createStore A function that receives the props and returns a core
* store object.
* @param props The props to pass to the createStore function.
*/
function useStore(createStore, props) {
	const [store, setStore] = React.useState(() => createStore(props));
	useSafeLayoutEffect(() => init(store), [store]);
	const useState = React.useCallback((keyOrSelector) => useStoreState(store, keyOrSelector), [store]);
	return [React.useMemo(() => ({
		...store,
		useState
	}), [store, useState]), useEvent(() => {
		setStore((store) => createStore({
			...props,
			...store.getState()
		}));
	})];
}
//#endregion
export { useStore, useStoreProps, useStoreState, useStoreStateObject };
