// @ariakit/react/cjs/__chunks/3FFOX2CW.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _4F2JVI6Tcjs = require('./4F2JVI6T.cjs');


var _ZBISR5IRcjs = require('./ZBISR5IR.cjs');



var _55MQXLFKcjs = require('./55MQXLFK.cjs');



var _SJZEGV7Acjs = require('./SJZEGV7A.cjs');




var _P52EZBZXcjs = require('./P52EZBZX.cjs');


var _AQHXJA6Mcjs = require('./AQHXJA6M.cjs');


var _KV7G7UO7cjs = require('./KV7G7UO7.cjs');



var _NQHLCEXFcjs = require('./NQHLCEXF.cjs');










var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');


var _BJQJHCU2cjs = require('./BJQJHCU2.cjs');

















var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/tab/tab-context.js
var ctx = _GF35MDH7cjs.createStoreContext.call(void 0, [_P52EZBZXcjs.CompositeContextProvider], [_P52EZBZXcjs.CompositeScopedContextProvider]);
var useTabContext = ctx.useContext;
var useTabScopedContext = ctx.useScopedContext;
var useTabProviderContext = ctx.useProviderContext;
var TabContextProvider = ctx.ContextProvider;
var TabScopedContextProvider = ctx.ScopedContextProvider;

// ../ariakit-react-components/dist/tab/tab.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useTab = _GF35MDH7cjs.createHook.call(void 0, function useTab2({ store, getItem: getItemProp, ...props }) {
  var _a;
  const context = useTabScopedContext();
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "Tab must be wrapped in a TabList component.");
  const defaultId = _GF35MDH7cjs.useId.call(void 0, );
  const id = props.id || defaultId;
  const dimmed = _GF35MDH7cjs.disabledFromProps.call(void 0, props);
  const getItem = _react.useCallback.call(void 0, (item) => {
    const nextItem = {
      ...item,
      dimmed
    };
    if (getItemProp) return getItemProp(nextItem);
    return nextItem;
  }, [dimmed, getItemProp]);
  const onClickProp = props.onClick;
  const onClick = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    store == null ? void 0 : store.setSelectedId(id);
  });
  const panelId = _P5CWMRCEcjs.useStoreState.call(void 0, store.panels, ["items"], () => {
    var _a2;
    return (_a2 = store.panel(id)) == null ? void 0 : _a2.id;
  });
  const shouldRegisterItem = defaultId ? props.shouldRegisterItem : false;
  const selected = _P5CWMRCEcjs.useStoreState.call(void 0, store, ["selectedId"], (state) => !!id && state.selectedId === id);
  const canRegisterComposedItem = _P5CWMRCEcjs.useStoreState.call(void 0, store.composite ? store : void 0, [
    "activeId",
    "items",
    "selectedId"
  ], (state) => {
    if (!state) return false;
    if (!id) return false;
    if (state.activeId === id) return true;
    if (state.selectedId !== id) return false;
    return !store.item(state.activeId);
  });
  const accessibleWhenDisabled = selected || ((_a = props.accessibleWhenDisabled) != null ? _a : true);
  if (_P5CWMRCEcjs.useStoreState.call(void 0, store.combobox || store.composite, "virtualFocus")) props = {
    ...props,
    tabIndex: -1
  };
  props = {
    role: "tab",
    "aria-selected": selected,
    "aria-controls": panelId || void 0,
    ...props,
    id,
    onClick
  };
  if (store.composite) {
    const defaultProps = {
      id,
      accessibleWhenDisabled,
      store: store.composite,
      shouldRegisterItem: canRegisterComposedItem && shouldRegisterItem,
      rowId: props.rowId,
      render: props.render
    };
    props = {
      ...props,
      render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _SJZEGV7Acjs.CompositeItem, {
        ...defaultProps,
        render: store.combobox && store.composite !== store.combobox ? /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _SJZEGV7Acjs.CompositeItem, {
          ...defaultProps,
          store: store.combobox
        }) : defaultProps.render
      })
    };
  }
  props = _SJZEGV7Acjs.useCompositeItem.call(void 0, {
    store,
    ...props,
    accessibleWhenDisabled,
    getItem,
    shouldRegisterItem
  });
  return props;
});
var Tab = _GF35MDH7cjs.memo.call(void 0, _GF35MDH7cjs.forwardRef.call(void 0, function Tab2(props) {
  const htmlProps = useTab(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
}));

// ../ariakit-react-components/dist/tab/tab-list.js

var TagName2 = "div";
var useTabList = _GF35MDH7cjs.createHook.call(void 0, function useTabList2({ store, ...props }) {
  const context = useTabProviderContext();
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "TabList must receive a `store` prop or be wrapped in a TabProvider component.");
  const orientation = _P5CWMRCEcjs.useStoreState.call(void 0, store, ["orientation"], (state) => state.orientation === "both" ? void 0 : state.orientation);
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, TabScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  if (store.composite) props = {
    focusable: false,
    ...props
  };
  props = {
    role: "tablist",
    "aria-orientation": orientation,
    ...props
  };
  props = _55MQXLFKcjs.useComposite.call(void 0, {
    store,
    ...props
  });
  return props;
});
var TabList = _GF35MDH7cjs.forwardRef.call(void 0, function TabList2(props) {
  const htmlProps = useTabList(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});

// ../ariakit-react-components/dist/tab/tab-panel.js


var TagName3 = "div";
var useTabPanel = _GF35MDH7cjs.createHook.call(void 0, function useTabPanel2({ store, unmountOnHide, tabId: tabIdProp, getItem: getItemProp, scrollRestoration, scrollElement, ...props }) {
  const context = useTabProviderContext();
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "TabPanel must receive a `store` prop or be wrapped in a TabProvider component.");
  const ref = _react.useRef.call(void 0, null);
  const id = _GF35MDH7cjs.useId.call(void 0, props.id);
  const tabId = _P5CWMRCEcjs.useStoreState.call(void 0, tabIdProp ? void 0 : store.panels, ["items"], () => {
    var _a;
    return tabIdProp || ((_a = store == null ? void 0 : store.panels.item(id)) == null ? void 0 : _a.tabId);
  });
  const disclosure = _NQHLCEXFcjs.useDisclosureStore.call(void 0, { open: _P5CWMRCEcjs.useStoreState.call(void 0, store, ["selectedId"], (state) => !!tabId && state.selectedId === tabId) });
  const mounted = _P5CWMRCEcjs.useStoreState.call(void 0, disclosure, "mounted");
  const scrollPositionRef = _react.useRef.call(void 0, null);
  const getScrollElement = _GF35MDH7cjs.useEvent.call(void 0, () => {
    const panelElement = ref.current;
    if (!panelElement) return null;
    if (!scrollElement) return panelElement;
    if (typeof scrollElement === "function") return scrollElement(panelElement);
    if ("current" in scrollElement) return scrollElement.current;
    return scrollElement;
  });
  _react.useEffect.call(void 0, () => {
    var _a, _b;
    if (!scrollRestoration) return;
    if (!mounted) return;
    const element = getScrollElement();
    if (!element) return;
    if (scrollRestoration === "reset") {
      element.scroll(0, 0);
      return;
    }
    if (!tabId) return;
    let scrollPositions = scrollPositionRef.current;
    if (!scrollPositions) {
      scrollPositions = /* @__PURE__ */ new Map();
      scrollPositionRef.current = scrollPositions;
    }
    const position = scrollPositions.get(tabId);
    element.scroll((_a = position == null ? void 0 : position.x) != null ? _a : 0, (_b = position == null ? void 0 : position.y) != null ? _b : 0);
    const onScroll = () => {
      scrollPositions.set(tabId, {
        x: element.scrollLeft,
        y: element.scrollTop
      });
    };
    element.addEventListener("scroll", onScroll);
    return () => {
      element.removeEventListener("scroll", onScroll);
    };
  }, [
    scrollRestoration,
    mounted,
    tabId,
    getScrollElement
  ]);
  const [hasTabbableChildren, setHasTabbableChildren] = _react.useState.call(void 0, false);
  _GF35MDH7cjs.useSafeLayoutEffect.call(void 0, () => {
    if (!mounted) return;
    const element = ref.current;
    if (!element) return;
    setHasTabbableChildren(!!_GF35MDH7cjs.getFirstTabbableIn.call(void 0, element));
  }, [mounted, tabId]);
  const getItem = _react.useCallback.call(void 0, (item) => {
    const nextItem = {
      ...item,
      id: id || item.id,
      tabId: tabIdProp
    };
    if (getItemProp) return getItemProp(nextItem);
    return nextItem;
  }, [
    id,
    tabIdProp,
    getItemProp
  ]);
  const onKeyDownProp = props.onKeyDown;
  const onKeyDown = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.defaultPrevented) return;
    if (!(store == null ? void 0 : store.composite)) return;
    const action = {
      ArrowLeft: store.previous,
      ArrowRight: store.next,
      Home: store.first,
      End: store.last
    }[event.key];
    if (!action) return;
    const { selectedId } = store.getState();
    const nextId = action({ activeId: selectedId });
    if (!nextId) return;
    event.preventDefault();
    store.move(nextId);
  });
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, TabScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  props = {
    role: "tabpanel",
    "aria-labelledby": props["aria-label"] != null ? void 0 : tabId || void 0,
    ...props,
    id,
    children: unmountOnHide && !mounted ? null : props.children,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref),
    onKeyDown
  };
  props = _BJQJHCU2cjs.useFocusable.call(void 0, {
    focusable: !store.composite && !hasTabbableChildren,
    ...props
  });
  props = _NQHLCEXFcjs.useDisclosureContent.call(void 0, {
    store: disclosure,
    ...props
  });
  props = _AQHXJA6Mcjs.useCollectionItem.call(void 0, {
    store: store.panels,
    ...props,
    getItem
  });
  return props;
});
var TabPanel = _GF35MDH7cjs.forwardRef.call(void 0, function TabPanel2(props) {
  const htmlProps = useTabPanel(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName3, htmlProps);
});

// ../ariakit-react-components/dist/tab/tab-store.js


// ../ariakit-components/dist/tab/tab-store.js
function getFocusedTab(items) {
  var _a, _b;
  const activeElement = (_b = (_a = items[0]) == null ? void 0 : _a.element) == null ? void 0 : _b.ownerDocument.activeElement;
  if (!activeElement) return;
  return items.find((item) => item.element === activeElement);
}
function getTabById(items, id) {
  if (id == null) return;
  return items.find((item) => item.id === id);
}
function isEnabledTab(item) {
  if (!item) return false;
  if (item.disabled) return false;
  if (item.dimmed) return false;
  return true;
}
function createPanelsStore() {
  const panels = _KV7G7UO7cjs.createCollectionStore.call(void 0, );
  const panelsByTabId = /* @__PURE__ */ new Map();
  _P5CWMRCEcjs.sync.call(void 0, panels, ["items"], (state) => {
    panelsByTabId.clear();
    for (const panel of state.items) {
      const { tabId } = panel;
      if (tabId == null) continue;
      if (panelsByTabId.has(tabId)) continue;
      panelsByTabId.set(tabId, panel);
    }
  });
  return {
    panels,
    panel: (tabId) => {
      if (tabId == null) return null;
      return panelsByTabId.get(tabId) || null;
    }
  };
}
function createTabStore({ composite: parentComposite, combobox, ...props } = {}) {
  const independentKeys = [
    "items",
    "renderedItems",
    "moves",
    "orientation",
    "virtualFocus",
    "includesBaseElement",
    "baseElement",
    "focusLoop",
    "focusShift",
    "focusWrap"
  ];
  const store = _P5CWMRCEcjs.mergeStore.call(void 0, props.store, _P5CWMRCEcjs.omit.call(void 0, parentComposite, independentKeys), _P5CWMRCEcjs.omit.call(void 0, combobox, independentKeys));
  const syncState = store == null ? void 0 : store.getState();
  const composite = _P52EZBZXcjs.createCompositeStore.call(void 0, {
    ...props,
    store,
    includesBaseElement: _GF35MDH7cjs.defaultValue.call(void 0, props.includesBaseElement, syncState == null ? void 0 : syncState.includesBaseElement, false),
    orientation: _GF35MDH7cjs.defaultValue.call(void 0, props.orientation, syncState == null ? void 0 : syncState.orientation, "horizontal"),
    focusLoop: _GF35MDH7cjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  });
  const { panels, panel } = createPanelsStore();
  const tab = _P5CWMRCEcjs.createStore.call(void 0, {
    ...composite.getState(),
    selectedId: _GF35MDH7cjs.defaultValue.call(void 0, props.selectedId, syncState == null ? void 0 : syncState.selectedId, props.defaultSelectedId),
    selectOnMove: _GF35MDH7cjs.defaultValue.call(void 0, props.selectOnMove, syncState == null ? void 0 : syncState.selectOnMove, true)
  }, composite, store);
  _P5CWMRCEcjs.setup.call(void 0, tab, () => _P5CWMRCEcjs.sync.call(void 0, tab, ["moves"], () => {
    const { activeId, selectOnMove } = tab.getState();
    if (!selectOnMove) return;
    if (!activeId) return;
    const tabItem = composite.item(activeId);
    if (!isEnabledTab(tabItem)) return;
    tab.setState("selectedId", tabItem.id);
  }));
  let pendingRestore = false;
  let restoredSelectedId;
  _P5CWMRCEcjs.setup.call(void 0, tab, () => {
    pendingRestore = false;
    return _P5CWMRCEcjs.batch.call(void 0, tab, ["selectedId"], (state, prev) => {
      if (pendingRestore) {
        pendingRestore = false;
        if (state.selectedId === restoredSelectedId) return;
      }
      if (parentComposite && state.selectedId === prev.selectedId) return;
      const { activeId, renderedItems } = tab.getState();
      if (activeId === state.selectedId) return;
      const focusedTab = getFocusedTab(renderedItems);
      const selectedTab = getTabById(renderedItems, state.selectedId);
      if (focusedTab && isEnabledTab(selectedTab)) {
        composite.move(selectedTab.id);
        return;
      }
      tab.setState("activeId", state.selectedId);
    });
  });
  _P5CWMRCEcjs.setup.call(void 0, tab, () => _P5CWMRCEcjs.sync.call(void 0, tab, ["selectedId", "renderedItems"], (state) => {
    if (state.selectedId !== void 0) return;
    const { activeId, renderedItems } = tab.getState();
    const tabItem = composite.item(activeId);
    if (isEnabledTab(tabItem)) tab.setState("selectedId", tabItem.id);
    else {
      const tabItem2 = renderedItems.find(isEnabledTab);
      tab.setState("selectedId", tabItem2 == null ? void 0 : tabItem2.id);
    }
  }));
  _P5CWMRCEcjs.setup.call(void 0, tab, () => _P5CWMRCEcjs.sync.call(void 0, tab, ["renderedItems"], (state) => {
    const tabs = state.renderedItems;
    if (!tabs.length) return;
    return _P5CWMRCEcjs.sync.call(void 0, panels, ["renderedItems"], (state2) => {
      const items = state2.renderedItems;
      if (!items.some((panel2) => !panel2.tabId)) return;
      items.forEach((panel2, i) => {
        if (panel2.tabId) return;
        const tabItem = tabs[i];
        if (!tabItem) return;
        panels.renderItem({
          ...panel2,
          tabId: tabItem.id
        });
      });
    });
  }));
  let selectedIdFromSelectedValue = null;
  _P5CWMRCEcjs.setup.call(void 0, tab, () => {
    const backupSelectedId = () => {
      selectedIdFromSelectedValue = tab.getState().selectedId;
    };
    const restoreSelectedId = () => {
      const { selectedId } = tab.getState();
      if (selectedId === selectedIdFromSelectedValue) return;
      pendingRestore = true;
      restoredSelectedId = selectedIdFromSelectedValue;
      tab.setState("selectedId", selectedIdFromSelectedValue);
    };
    if (parentComposite && "setSelectElement" in parentComposite) return _GF35MDH7cjs.chain.call(void 0, _P5CWMRCEcjs.sync.call(void 0, parentComposite, ["value"], backupSelectedId), _P5CWMRCEcjs.sync.call(void 0, parentComposite, ["mounted"], restoreSelectedId));
    if (!combobox) return;
    return _GF35MDH7cjs.chain.call(void 0, _P5CWMRCEcjs.sync.call(void 0, combobox, ["selectedValue"], backupSelectedId), _P5CWMRCEcjs.sync.call(void 0, combobox, ["mounted"], restoreSelectedId));
  });
  return {
    ...composite,
    ...tab,
    panels,
    panel,
    setSelectedId: (id) => tab.setState("selectedId", id),
    select: (id) => {
      tab.setState("selectedId", id);
      composite.move(id);
    }
  };
}

// ../ariakit-react-components/dist/tab/tab-store.js
function useTabStoreProps(store, update, props) {
  _GF35MDH7cjs.useUpdateEffect.call(void 0, update, [props.composite, props.combobox]);
  const compositeStore = _55MQXLFKcjs.useCompositeStoreProps.call(void 0, store, update, props);
  _P5CWMRCEcjs.useStoreProps.call(void 0, compositeStore, props, "selectedId", "setSelectedId");
  _P5CWMRCEcjs.useStoreProps.call(void 0, compositeStore, props, "selectOnMove");
  const [panels, updatePanels] = _P5CWMRCEcjs.useStore.call(void 0, () => compositeStore.panels, {});
  _GF35MDH7cjs.useUpdateEffect.call(void 0, updatePanels, [compositeStore, updatePanels]);
  return Object.assign(_react.useMemo.call(void 0, () => ({
    ...compositeStore,
    panels
  }), [compositeStore, panels]), {
    composite: props.composite,
    combobox: props.combobox
  });
}
function useTabStore(props = {}) {
  const combobox = _ZBISR5IRcjs.useComboboxContext.call(void 0, );
  const composite = _4F2JVI6Tcjs.useSelectContext.call(void 0, ) || combobox;
  const storeProps = {
    ...props,
    composite: props.composite !== void 0 ? props.composite : composite,
    combobox: props.combobox !== void 0 ? props.combobox : combobox
  };
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, createTabStore, storeProps);
  return useTabStoreProps(store, update, storeProps);
}

// ../ariakit-react-components/dist/tab/tab-provider.js

function TabProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, TabContextProvider, {
    value: useTabStore(props),
    children: props.children
  });
}








exports.useTabContext = useTabContext; exports.Tab = Tab; exports.TabList = TabList; exports.TabPanel = TabPanel; exports.useTabStore = useTabStore; exports.TabProvider = TabProvider;
