// @ariakit/react/cjs/__chunks/3LMRVZ7A.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _NRHBFZM7cjs = require('./NRHBFZM7.cjs');


var _JZGTMT3Qcjs = require('./JZGTMT3Q.cjs');




var _KAU3JWAXcjs = require('./KAU3JWAX.cjs');


var _NBWQLKS5cjs = require('./NBWQLKS5.cjs');


var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');








var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/popover/popover-disclosure.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var usePopoverDisclosure = _GF35MDH7cjs.createHook.call(void 0, function usePopoverDisclosure2({ store, ...props }) {
  const context = _KAU3JWAXcjs.usePopoverProviderContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "PopoverDisclosure must receive a `store` prop or be wrapped in a PopoverProvider component.");
  const onClickProp = props.onClick;
  const onClick = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    store == null ? void 0 : store.setAnchorElement(event.currentTarget);
    onClickProp == null ? void 0 : onClickProp(event);
  });
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _KAU3JWAXcjs.PopoverScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  props = {
    ...props,
    onClick
  };
  props = _NRHBFZM7cjs.usePopoverAnchor.call(void 0, {
    store,
    ...props
  });
  props = _NBWQLKS5cjs.useDialogDisclosure.call(void 0, {
    store,
    ...props
  });
  return props;
});
var PopoverDisclosure = _GF35MDH7cjs.forwardRef.call(void 0, function PopoverDisclosure2(props) {
  const htmlProps = usePopoverDisclosure(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/popover/popover-disclosure-arrow.js
var _react = require('react');

var TagName2 = "span";
var pointsMap = {
  top: "4,10 8,6 12,10",
  right: "6,4 10,8 6,12",
  bottom: "4,6 8,10 12,6",
  left: "10,4 6,8 10,12"
};
var usePopoverDisclosureArrow = _GF35MDH7cjs.createHook.call(void 0, function usePopoverDisclosureArrow2({ store, placement, ...props }) {
  const context = _KAU3JWAXcjs.usePopoverContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "PopoverDisclosureArrow must be wrapped in a PopoverDisclosure component.");
  const dir = _JZGTMT3Qcjs.getBasePlacement.call(void 0, _P5CWMRCEcjs.useStoreState.call(void 0, store, ["placement"], (state) => placement || state.placement));
  const points = pointsMap[dir];
  props = {
    children: _react.useMemo.call(void 0, () => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "svg", {
      display: "block",
      fill: "none",
      stroke: "currentColor",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.5,
      viewBox: "0 0 16 16",
      height: "1em",
      width: "1em",
      children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points })
    }), [points]),
    "aria-hidden": true,
    ...props,
    style: {
      width: "1em",
      height: "1em",
      pointerEvents: "none",
      ...props.style
    }
  };
  return _GF35MDH7cjs.removeUndefinedValues.call(void 0, props);
});
var PopoverDisclosureArrow = _GF35MDH7cjs.forwardRef.call(void 0, function PopoverDisclosureArrow2(props) {
  const htmlProps = usePopoverDisclosureArrow(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});






exports.usePopoverDisclosure = usePopoverDisclosure; exports.PopoverDisclosure = PopoverDisclosure; exports.usePopoverDisclosureArrow = usePopoverDisclosureArrow; exports.PopoverDisclosureArrow = PopoverDisclosureArrow;
