// @ariakit/react/cjs/__chunks/3PZT7FSK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _N6COOBTRcjs = require('./N6COOBTR.cjs');



var _NQHLCEXFcjs = require('./NQHLCEXF.cjs');


var _ZQUU53RKcjs = require('./ZQUU53RK.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');







var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/dialog/dialog-dismiss.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useDialogDismiss = _GF35MDH7cjs.createHook.call(void 0, function useDialogDismiss2({ store, ...props }) {
  const context = _NQHLCEXFcjs.useDialogScopedContext.call(void 0, );
  store = store || context;
  const onClickProp = props.onClick;
  const onClick = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    store == null ? void 0 : store.hide();
  });
  props = {
    "data-dialog-dismiss": "",
    children: _react.useMemo.call(void 0, () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, "svg", {
      "aria-label": "Dismiss popup",
      display: "block",
      fill: "none",
      stroke: "currentColor",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.5,
      viewBox: "0 0 16 16",
      height: "1em",
      width: "1em",
      children: [/* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", {
        x1: "4",
        y1: "4",
        x2: "12",
        y2: "12"
      }), /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", {
        x1: "4",
        y1: "12",
        x2: "12",
        y2: "4"
      })]
    }), []),
    ...props,
    onClick
  };
  props = _ZQUU53RKcjs.useButton.call(void 0, props);
  return props;
});
var DialogDismiss = _GF35MDH7cjs.forwardRef.call(void 0, function DialogDismiss2(props) {
  const htmlProps = useDialogDismiss(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/dialog/dialog-heading.js

var TagName2 = "h1";
var useDialogHeading = _GF35MDH7cjs.createHook.call(void 0, function useDialogHeading2({ store, ...props }) {
  const setHeadingId = _react.useContext.call(void 0, _NQHLCEXFcjs.DialogHeadingContext);
  const id = _GF35MDH7cjs.useId.call(void 0, props.id);
  _GF35MDH7cjs.useSafeLayoutEffect.call(void 0, () => {
    setHeadingId == null ? void 0 : setHeadingId(id);
    return () => setHeadingId == null ? void 0 : setHeadingId(void 0);
  }, [setHeadingId, id]);
  props = {
    ...props,
    id
  };
  props = _N6COOBTRcjs.useHeading.call(void 0, props);
  return props;
});
var DialogHeading = _GF35MDH7cjs.forwardRef.call(void 0, function DialogHeading2(props) {
  const htmlProps = useDialogHeading(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});






exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss; exports.useDialogHeading = useDialogHeading; exports.DialogHeading = DialogHeading;
