// @ariakit/react/cjs/__chunks/4F2JVI6T.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _KAU3JWAXcjs = require('./KAU3JWAX.cjs');



var _P52EZBZXcjs = require('./P52EZBZX.cjs');


var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/select/select-context.js
var _react = require('react');
var ctx = _GF35MDH7cjs.createStoreContext.call(void 0, [_KAU3JWAXcjs.PopoverContextProvider, _P52EZBZXcjs.CompositeContextProvider], [_KAU3JWAXcjs.PopoverScopedContextProvider, _P52EZBZXcjs.CompositeScopedContextProvider]);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
