// @ariakit/react/cjs/__chunks/55MQXLFK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});








var _P52EZBZXcjs = require('./P52EZBZX.cjs');


var _KV7G7UO7cjs = require('./KV7G7UO7.cjs');




var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');


var _BJQJHCU2cjs = require('./BJQJHCU2.cjs');






















var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/composite/composite.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
function isGrid(items) {
  return items.some((item) => !!item.rowId);
}
function isPrintableKey(event) {
  const target = event.target;
  if (target && !_GF35MDH7cjs.isTextField.call(void 0, target)) return false;
  return event.key.length === 1 && !event.ctrlKey && !event.metaKey;
}
function isModifierKey(event) {
  return event.key === "Shift" || event.key === "Control" || event.key === "Alt" || event.key === "Meta";
}
function useKeyboardEventProxy(store, onKeyboardEvent, previousElementRef) {
  return _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    var _a;
    onKeyboardEvent == null ? void 0 : onKeyboardEvent(event);
    if (event.defaultPrevented) return;
    if (event.isPropagationStopped()) return;
    if (!_GF35MDH7cjs.isSelfTarget.call(void 0, event)) return;
    if (isModifierKey(event)) return;
    if (isPrintableKey(event)) return;
    const activeElement = (_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, store.getState().activeId)) == null ? void 0 : _a.element;
    if (!activeElement) return;
    const { view, ...eventInit } = event;
    if (activeElement !== (previousElementRef == null ? void 0 : previousElementRef.current)) activeElement.focus();
    if (!_GF35MDH7cjs.fireKeyboardEvent.call(void 0, activeElement, event.type, eventInit)) event.preventDefault();
    if (event.currentTarget.contains(activeElement)) event.stopPropagation();
  });
}
function findFirstEnabledItemInTheLastRow(items) {
  return _P52EZBZXcjs.findFirstEnabledItem.call(void 0, _GF35MDH7cjs.flatten2DArray.call(void 0, _GF35MDH7cjs.reverseArray.call(void 0, _P52EZBZXcjs.groupItemsByRows.call(void 0, items))));
}
function withBaseScrollPreserved(store, callback) {
  const { virtualFocus, baseElement } = store.getState();
  if (!virtualFocus || !baseElement || !_GF35MDH7cjs.isTextField.call(void 0, baseElement)) {
    callback();
    return;
  }
  const savedScrollLeft = baseElement.scrollLeft;
  const savedScrollTop = baseElement.scrollTop;
  callback();
  baseElement.scrollLeft = savedScrollLeft;
  baseElement.scrollTop = savedScrollTop;
}
function useScheduleFocus(store) {
  const [scheduled, setScheduled] = _react.useState.call(void 0, false);
  const schedule = _react.useCallback.call(void 0, () => setScheduled(true), []);
  const activeItem = _P5CWMRCEcjs.useStoreState.call(void 0, store, scheduled ? ["activeId", "items"] : [], (state) => scheduled ? _P52EZBZXcjs.getEnabledItem.call(void 0, store, state.activeId) : null);
  _react.useEffect.call(void 0, () => {
    const activeElement = activeItem == null ? void 0 : activeItem.element;
    if (!scheduled) return;
    if (!activeElement) return;
    setScheduled(false);
    withBaseScrollPreserved(store, () => {
      activeElement.focus({ preventScroll: true });
    });
  }, [
    store,
    activeItem,
    scheduled
  ]);
  return schedule;
}
var CompositeFocusOnMove = _GF35MDH7cjs.memo.call(void 0, function CompositeFocusOnMove2({ store, focusOnMove, previousElementRef }) {
  const moves = _P5CWMRCEcjs.useStoreState.call(void 0, store, "moves");
  const baseElement = _P5CWMRCEcjs.useStoreState.call(void 0, store, "baseElement");
  _react.useEffect.call(void 0, () => {
    var _a;
    if (!moves) return;
    if (!focusOnMove) return;
    const { activeId } = store.getState();
    const itemElement = (_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, activeId)) == null ? void 0 : _a.element;
    if (!itemElement) return;
    withBaseScrollPreserved(store, () => _GF35MDH7cjs.focusIntoView.call(void 0, itemElement));
  }, [
    store,
    moves,
    focusOnMove
  ]);
  _GF35MDH7cjs.useSafeLayoutEffect.call(void 0, () => {
    if (!moves) return;
    if (!baseElement) return;
    const { activeId } = store.getState();
    if (!(activeId === null)) return;
    const previousElement = previousElementRef.current;
    previousElementRef.current = null;
    if (previousElement) _GF35MDH7cjs.fireBlurEvent.call(void 0, previousElement, { relatedTarget: baseElement });
    if (!_GF35MDH7cjs.hasFocus.call(void 0, baseElement)) baseElement.focus();
  }, [
    store,
    moves,
    baseElement
  ]);
  return null;
});
var useComposite = _GF35MDH7cjs.createHook.call(void 0, function useComposite2({ store, composite = true, focusOnMove = composite, moveOnKeyPress = true, ...props }) {
  const context = _P52EZBZXcjs.useCompositeProviderContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "Composite must receive a `store` prop or be wrapped in a CompositeProvider component.");
  const ref = _react.useRef.call(void 0, null);
  const previousElementRef = _react.useRef.call(void 0, null);
  const scheduleFocus = useScheduleFocus(store);
  const [, setBaseElement] = _GF35MDH7cjs.useTransactionState.call(void 0, composite ? store.setBaseElement : null);
  const virtualFocus = _P5CWMRCEcjs.useStoreState.call(void 0, store, "virtualFocus");
  const activeId = _P5CWMRCEcjs.useStoreState.call(void 0, store, composite && virtualFocus ? ["activeId"] : [], (state) => composite && virtualFocus ? state.activeId : null);
  _GF35MDH7cjs.useSafeLayoutEffect.call(void 0, () => {
    var _a;
    if (!store) return;
    if (!composite) return;
    if (!virtualFocus) return;
    const previousElement = previousElementRef.current;
    previousElementRef.current = null;
    if (!previousElement) return;
    const relatedTarget = ((_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, activeId)) == null ? void 0 : _a.element) || _GF35MDH7cjs.getActiveElement.call(void 0, previousElement);
    if (relatedTarget === previousElement) return;
    _GF35MDH7cjs.fireBlurEvent.call(void 0, previousElement, { relatedTarget });
  }, [
    store,
    activeId,
    virtualFocus,
    composite
  ]);
  const onKeyDownCapture = useKeyboardEventProxy(store, props.onKeyDownCapture, previousElementRef);
  const onKeyUpCapture = useKeyboardEventProxy(store, props.onKeyUpCapture, previousElementRef);
  const onFocusCaptureProp = props.onFocusCapture;
  const onFocusCapture = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onFocusCaptureProp == null ? void 0 : onFocusCaptureProp(event);
    if (event.defaultPrevented) return;
    if (!store) return;
    const { virtualFocus: virtualFocus2 } = store.getState();
    if (!virtualFocus2) return;
    const previousActiveElement = event.relatedTarget;
    const isSilentlyFocused = _P52EZBZXcjs.silentlyFocused.call(void 0, event.currentTarget);
    if (_GF35MDH7cjs.isSelfTarget.call(void 0, event) && isSilentlyFocused) {
      event.stopPropagation();
      previousElementRef.current = previousActiveElement;
    }
  });
  const onFocusProp = props.onFocus;
  const onFocus = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onFocusProp == null ? void 0 : onFocusProp(event);
    if (event.defaultPrevented) return;
    if (!composite) return;
    if (!store) return;
    const { relatedTarget } = event;
    const { virtualFocus: virtualFocus2 } = store.getState();
    if (virtualFocus2) {
      if (_GF35MDH7cjs.isSelfTarget.call(void 0, event) && !_P52EZBZXcjs.isItem.call(void 0, store, relatedTarget)) queueMicrotask(scheduleFocus);
    } else if (_GF35MDH7cjs.isSelfTarget.call(void 0, event)) store.setActiveId(null);
  });
  const onBlurCaptureProp = props.onBlurCapture;
  const onBlurCapture = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    var _a;
    onBlurCaptureProp == null ? void 0 : onBlurCaptureProp(event);
    if (event.defaultPrevented) return;
    if (!store) return;
    const { virtualFocus: virtualFocus2, activeId: activeId2 } = store.getState();
    if (!virtualFocus2) return;
    const activeElement = (_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, activeId2)) == null ? void 0 : _a.element;
    const nextActiveElement = event.relatedTarget;
    const nextActiveElementIsItem = _P52EZBZXcjs.isItem.call(void 0, store, nextActiveElement);
    const previousElement = previousElementRef.current;
    previousElementRef.current = null;
    if (_GF35MDH7cjs.isSelfTarget.call(void 0, event) && nextActiveElementIsItem) {
      if (nextActiveElement === activeElement) {
        if (previousElement && previousElement !== nextActiveElement) _GF35MDH7cjs.fireBlurEvent.call(void 0, previousElement, event);
      } else if (activeElement) _GF35MDH7cjs.fireBlurEvent.call(void 0, activeElement, event);
      else if (previousElement) _GF35MDH7cjs.fireBlurEvent.call(void 0, previousElement, event);
      event.stopPropagation();
    } else if (!_P52EZBZXcjs.isItem.call(void 0, store, event.target) && activeElement) _GF35MDH7cjs.fireBlurEvent.call(void 0, activeElement, event);
  });
  const onKeyDownProp = props.onKeyDown;
  const moveOnKeyPressProp = _GF35MDH7cjs.useBooleanEvent.call(void 0, moveOnKeyPress);
  const onKeyDown = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    var _a, _b;
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.nativeEvent.isComposing) return;
    if (event.defaultPrevented) return;
    if (!store) return;
    if (!_GF35MDH7cjs.isSelfTarget.call(void 0, event)) return;
    const { orientation, renderedItems, activeId: activeId2, rtl } = store.getState();
    if ((_b = (_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, activeId2)) == null ? void 0 : _a.element) == null ? void 0 : _b.isConnected) return;
    const isVertical = orientation !== "horizontal";
    const isHorizontal = orientation !== "vertical";
    const grid = isGrid(renderedItems);
    if ((event.key === "ArrowLeft" || event.key === "ArrowRight" || event.key === "Home" || event.key === "End") && _GF35MDH7cjs.isTextField.call(void 0, event.currentTarget)) return;
    const up = () => {
      var _a2;
      if (grid) return (_a2 = findFirstEnabledItemInTheLastRow(renderedItems)) == null ? void 0 : _a2.id;
      return store == null ? void 0 : store.last();
    };
    const action = {
      ArrowUp: (grid || isVertical) && up,
      ArrowRight: (grid || isHorizontal) && (rtl ? store.last : store.first),
      ArrowDown: (grid || isVertical) && store.first,
      ArrowLeft: (grid || isHorizontal) && (rtl ? store.first : store.last),
      Home: store.first,
      End: store.last,
      PageUp: store.first,
      PageDown: store.last
    }[event.key];
    if (action) {
      const id = action();
      if (id !== void 0) {
        if (!moveOnKeyPressProp(event)) return;
        event.preventDefault();
        store.move(id);
      }
    }
  });
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _P52EZBZXcjs.CompositeScopedContextProvider, {
    value: store,
    children: [element, composite && /* @__PURE__ */ _jsxruntime.jsx.call(void 0, CompositeFocusOnMove, {
      store,
      focusOnMove,
      previousElementRef
    })]
  }), [
    store,
    composite,
    focusOnMove
  ]);
  props = {
    "aria-activedescendant": _P5CWMRCEcjs.useStoreState.call(void 0, store, composite && virtualFocus ? ["items"] : [], () => {
      var _a;
      if (!store) return;
      if (!composite) return;
      if (!virtualFocus) return;
      return (_a = _P52EZBZXcjs.getEnabledItem.call(void 0, store, activeId)) == null ? void 0 : _a.id;
    }),
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, setBaseElement, props.ref),
    onKeyDownCapture,
    onKeyUpCapture,
    onFocusCapture,
    onFocus,
    onBlurCapture,
    onKeyDown
  };
  props = _BJQJHCU2cjs.useFocusable.call(void 0, {
    focusable: _P5CWMRCEcjs.useStoreState.call(void 0, store, composite && !virtualFocus ? ["activeId"] : [], (state) => composite && (virtualFocus || state.activeId === null)),
    ...props
  });
  return props;
});
var Composite = _GF35MDH7cjs.forwardRef.call(void 0, function Composite2(props) {
  const htmlProps = useComposite(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/composite/composite-store.js
function useCompositeStoreOptions(props) {
  return {
    id: _GF35MDH7cjs.useId.call(void 0, props.id),
    ...props
  };
}
function useCompositeStoreProps(store, update, props) {
  store = _KV7G7UO7cjs.useCollectionStoreProps.call(void 0, store, update, props);
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "activeId", "setActiveId");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "includesBaseElement");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "virtualFocus");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "orientation");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "rtl");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "focusLoop");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "focusWrap");
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "focusShift");
  return store;
}
function useCompositeStore(props = {}) {
  props = useCompositeStoreOptions(props);
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, _P52EZBZXcjs.createCompositeStore, props);
  return useCompositeStoreProps(store, update, props);
}







exports.useComposite = useComposite; exports.Composite = Composite; exports.useCompositeStoreOptions = useCompositeStoreOptions; exports.useCompositeStoreProps = useCompositeStoreProps; exports.useCompositeStore = useCompositeStore;
