// @ariakit/react/cjs/__chunks/6F6JBC2R.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _2WVP5ZDEcjs = require('./2WVP5ZDE.cjs');




var _55MQXLFKcjs = require('./55MQXLFK.cjs');


var _P52EZBZXcjs = require('./P52EZBZX.cjs');




var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');









var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/radio/radio-group.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
function isCheckedRadio(element) {
  if (!element) return false;
  if (element.tagName === "INPUT") {
    const { type, checked } = element;
    if (type === "radio") return checked;
  }
  if (element.getAttribute("role") !== "radio") return false;
  return element.getAttribute("aria-checked") === "true";
}
function getCheckedRadioId(store) {
  var _a;
  const { renderedItems } = store.getState();
  return (_a = renderedItems.find((item) => isCheckedRadio(item.element))) == null ? void 0 : _a.id;
}
var useRadioGroup = _GF35MDH7cjs.createHook.call(void 0, function useRadioGroup2({ store, ...props }) {
  const context = _2WVP5ZDEcjs.useRadioProviderContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "RadioGroup must receive a `store` prop or be wrapped in a RadioProvider component.");
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _2WVP5ZDEcjs.RadioScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  const onBlurCaptureProp = props.onBlurCapture;
  const onBlurCapture = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onBlurCaptureProp == null ? void 0 : onBlurCaptureProp(event);
    if (event.defaultPrevented) return;
    if (!_GF35MDH7cjs.isFocusEventOutside.call(void 0, event)) return;
    const checkedId = getCheckedRadioId(store);
    if (!checkedId) return;
    store.setActiveId(checkedId);
  });
  props = {
    role: "radiogroup",
    ...props,
    onBlurCapture
  };
  props = _55MQXLFKcjs.useComposite.call(void 0, {
    store,
    ...props
  });
  return props;
});
var RadioGroup = _GF35MDH7cjs.forwardRef.call(void 0, function RadioGroup2(props) {
  const htmlProps = useRadioGroup(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-components/dist/radio/radio-store.js
function createRadioStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const composite = _P52EZBZXcjs.createCompositeStore.call(void 0, {
    ...props,
    focusLoop: _GF35MDH7cjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  });
  const radio = _P5CWMRCEcjs.createStore.call(void 0, {
    ...composite.getState(),
    value: _GF35MDH7cjs.defaultValue.call(void 0, props.value, syncState == null ? void 0 : syncState.value, props.defaultValue, null)
  }, composite, props.store);
  return {
    ...composite,
    ...radio,
    setValue: (value) => radio.setState("value", value)
  };
}

// ../ariakit-react-components/dist/radio/radio-store.js
function useRadioStoreProps(store, update, props) {
  store = _55MQXLFKcjs.useCompositeStoreProps.call(void 0, store, update, props);
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "value", "setValue");
  return store;
}
function useRadioStore(props = {}) {
  props = _55MQXLFKcjs.useCompositeStoreOptions.call(void 0, props);
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, createRadioStore, props);
  return useRadioStoreProps(store, update, props);
}

// ../ariakit-react-components/dist/radio/radio-provider.js

function RadioProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _2WVP5ZDEcjs.RadioContextProvider, {
    value: useRadioStore(props),
    children: props.children
  });
}





exports.RadioGroup = RadioGroup; exports.useRadioStore = useRadioStore; exports.RadioProvider = RadioProvider;
