// @ariakit/react/cjs/__chunks/AQHXJA6M.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _KV7G7UO7cjs = require('./KV7G7UO7.cjs');








var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/collection/collection-item.js
var _react = require('react');
var TagName = "div";
var useCollectionItem = _GF35MDH7cjs.createHook.call(void 0, function useCollectionItem2({ store, shouldRegisterItem = true, getItem = _GF35MDH7cjs.identity, element, ...props }) {
  const context = _KV7G7UO7cjs.useCollectionContext.call(void 0, );
  store = store || context;
  const id = _GF35MDH7cjs.useId.call(void 0, props.id);
  const ref = _react.useRef.call(void 0, element);
  _react.useEffect.call(void 0, () => {
    const element2 = ref.current;
    if (!id) return;
    if (!element2) return;
    if (!shouldRegisterItem) return;
    const item = getItem({
      id,
      element: element2
    });
    return store == null ? void 0 : store.renderItem(item);
  }, [
    id,
    shouldRegisterItem,
    getItem,
    store
  ]);
  props = {
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  return _GF35MDH7cjs.removeUndefinedValues.call(void 0, props);
});
var CollectionItem = _GF35MDH7cjs.forwardRef.call(void 0, function CollectionItem2(props) {
  const htmlProps = useCollectionItem(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useCollectionItem = useCollectionItem; exports.CollectionItem = CollectionItem;
