// @ariakit/react/cjs/__chunks/ERZLFUVI.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/visually-hidden/visually-hidden.js
var TagName = "span";
function getVisuallyHiddenStyle(style) {
  return {
    borderWidth: 0,
    clipPath: "inset(50%)",
    height: "1px",
    margin: "-1px",
    overflow: "hidden",
    padding: 0,
    position: "absolute",
    whiteSpace: "nowrap",
    width: "1px",
    ...style
  };
}
var useVisuallyHidden = _GF35MDH7cjs.createHook.call(void 0, function useVisuallyHidden2(props) {
  props = {
    ...props,
    style: getVisuallyHiddenStyle(props.style)
  };
  return props;
});
var VisuallyHidden = _GF35MDH7cjs.forwardRef.call(void 0, function VisuallyHidden2(props) {
  const htmlProps = useVisuallyHidden(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});





exports.getVisuallyHiddenStyle = getVisuallyHiddenStyle; exports.useVisuallyHidden = useVisuallyHidden; exports.VisuallyHidden = VisuallyHidden;
