// @ariakit/react/cjs/__chunks/FQUMO3QY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _P52EZBZXcjs = require('./P52EZBZX.cjs');












var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/composite/composite-typeahead.js
var TagName = "div";
var typeaheadStates = /* @__PURE__ */ new WeakMap();
function getTypeaheadState(store) {
  let typeaheadState = typeaheadStates.get(store);
  if (!typeaheadState) {
    typeaheadState = {
      chars: "",
      cleanupTimeout: 0
    };
    typeaheadStates.set(store, typeaheadState);
  }
  return typeaheadState;
}
function clearChars(typeaheadState) {
  typeaheadState.chars = "";
}
function isValidTypeaheadEvent(event, typeaheadState) {
  const target = event.target;
  if (target && _GF35MDH7cjs.isTextField.call(void 0, target)) return false;
  if (event.key === " " && typeaheadState.chars.length) return true;
  return event.key.length === 1 && !event.ctrlKey && !event.altKey && !event.metaKey && /^[\p{Letter}\p{Number}]$/u.test(event.key);
}
function isSelfTargetOrItem(event, items) {
  if (_GF35MDH7cjs.isSelfTarget.call(void 0, event)) return true;
  const target = event.target;
  if (!target) return false;
  return items.some((item) => item.element === target);
}
function getEnabledItems(items) {
  return items.filter((item) => !item.disabled);
}
function itemTextStartsWith(item, text) {
  var _a;
  const itemText = ((_a = item.element) == null ? void 0 : _a.textContent) || item.children || "value" in item && item.value;
  if (!itemText) return false;
  return _GF35MDH7cjs.normalizeString.call(void 0, itemText).trim().toLowerCase().startsWith(text.toLowerCase());
}
function getSameInitialItems({ activeId, char, items, typeaheadState }) {
  if (!activeId) return items;
  const activeItem = items.find((item) => item.id === activeId);
  if (!activeItem) return items;
  if (!itemTextStartsWith(activeItem, char)) return items;
  const { chars } = typeaheadState;
  if (chars !== char && itemTextStartsWith(activeItem, chars)) return items;
  typeaheadState.chars = char;
  return _P52EZBZXcjs.flipItems.call(void 0, items.filter((item) => itemTextStartsWith(item, char)), activeId).filter((item) => item.id !== activeId);
}
var useCompositeTypeahead = _GF35MDH7cjs.createHook.call(void 0, function useCompositeTypeahead2({ store, typeahead = true, ...props }) {
  const context = _P52EZBZXcjs.useCompositeScopedContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "CompositeTypeahead must be a Composite component");
  const onKeyDownCaptureProp = props.onKeyDownCapture;
  const onKeyDownCapture = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onKeyDownCaptureProp == null ? void 0 : onKeyDownCaptureProp(event);
    if (event.defaultPrevented) return;
    if (!typeahead) return;
    if (!store) return;
    const typeaheadState = getTypeaheadState(store);
    if (!isValidTypeaheadEvent(event, typeaheadState)) return clearChars(typeaheadState);
    const { renderedItems, items, activeId, id } = store.getState();
    let enabledItems = getEnabledItems(items.length > renderedItems.length ? items : renderedItems);
    const document = _GF35MDH7cjs.getDocument.call(void 0, event.currentTarget);
    const selector = `[data-offscreen-id="${id}"]`;
    const offscreenItems = document.querySelectorAll(selector);
    for (const element of offscreenItems) {
      if (element.hasAttribute("data-disabled")) continue;
      if ("disabled" in element && !!element.disabled) continue;
      enabledItems.push({
        id: element.id,
        element
      });
    }
    if (offscreenItems.length) enabledItems = _GF35MDH7cjs.sortBasedOnDOMPosition.call(void 0, enabledItems, (i) => i.element);
    if (!isSelfTargetOrItem(event, enabledItems)) return clearChars(typeaheadState);
    event.preventDefault();
    window.clearTimeout(typeaheadState.cleanupTimeout);
    typeaheadState.cleanupTimeout = window.setTimeout(() => {
      clearChars(typeaheadState);
    }, 500);
    const char = event.key.toLowerCase();
    typeaheadState.chars += char;
    enabledItems = getSameInitialItems({
      activeId,
      char,
      items: enabledItems,
      typeaheadState
    });
    const item = enabledItems.find((item2) => itemTextStartsWith(item2, typeaheadState.chars));
    if (item) store.move(item.id);
    else clearChars(typeaheadState);
  });
  props = {
    ...props,
    onKeyDownCapture
  };
  return _GF35MDH7cjs.removeUndefinedValues.call(void 0, props);
});
var CompositeTypeahead = _GF35MDH7cjs.forwardRef.call(void 0, function CompositeTypeahead2(props) {
  const htmlProps = useCompositeTypeahead(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useCompositeTypeahead = useCompositeTypeahead; exports.CompositeTypeahead = CompositeTypeahead;
