// @ariakit/react/cjs/__chunks/KV7G7UO7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});







var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');







var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/collection/collection-context.js
var ctx = _GF35MDH7cjs.createStoreContext.call(void 0, );
var useCollectionContext = ctx.useContext;
var useCollectionScopedContext = ctx.useScopedContext;
var useCollectionProviderContext = ctx.useProviderContext;
var CollectionContextProvider = ctx.ContextProvider;
var CollectionScopedContextProvider = ctx.ScopedContextProvider;

// ../ariakit-components/dist/collection/collection-store.js
function getCommonParent(items) {
  var _a, _b;
  const firstItem = items.find((item) => !!item.element);
  const lastElement = (_a = [...items].reverse().find((item) => !!item.element)) == null ? void 0 : _a.element;
  let parentElement = (_b = firstItem == null ? void 0 : firstItem.element) == null ? void 0 : _b.parentElement;
  if (!lastElement) return _GF35MDH7cjs.getDocument.call(void 0, parentElement).body;
  while (parentElement) {
    if (parentElement.contains(lastElement)) return parentElement;
    parentElement = parentElement.parentElement;
  }
  return _GF35MDH7cjs.getDocument.call(void 0, parentElement).body;
}
function getPrivateStore(store) {
  return store == null ? void 0 : store.__unstablePrivateStore;
}
var itemIdCacheThreshold = 64;
function getCachedItemIds(items, cache) {
  if (items.length < itemIdCacheThreshold) {
    if (cache.ids) {
      cache.items = void 0;
      cache.ids = void 0;
    }
    return;
  }
  if (cache.items === items) return cache.ids;
  const ids = /* @__PURE__ */ new Set();
  for (const item of items) ids.add(item.id);
  cache.items = items;
  cache.ids = ids;
  return ids;
}
function setCachedItemIds(items, cache, ids) {
  if (items.length < itemIdCacheThreshold) {
    if (cache.ids) {
      cache.items = void 0;
      cache.ids = void 0;
    }
    return;
  }
  if (!ids) {
    ids = /* @__PURE__ */ new Set();
    for (const item of items) ids.add(item.id);
  }
  cache.items = items;
  cache.ids = ids;
}
function createCollectionStore(props = {}) {
  var _a;
  _P5CWMRCEcjs.throwOnConflictingProps.call(void 0, props, props.store);
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const items = _GF35MDH7cjs.defaultValue.call(void 0, props.items, syncState == null ? void 0 : syncState.items, props.defaultItems, []);
  const itemsMap = new Map(items.map((item) => [item.id, item]));
  const itemIdCache = {};
  const renderedItemIdCache = {};
  const initialState = {
    items,
    renderedItems: _GF35MDH7cjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.renderedItems, [])
  };
  const syncPrivateStore = getPrivateStore(props.store);
  const privateStore = _P5CWMRCEcjs.createStore.call(void 0, {
    items,
    renderedItems: initialState.renderedItems
  }, syncPrivateStore);
  const collection = _P5CWMRCEcjs.createStore.call(void 0, initialState, props.store);
  const sortItems = (renderedItems) => {
    const sortedItems = _GF35MDH7cjs.sortBasedOnDOMPosition.call(void 0, renderedItems, (i) => i.element);
    privateStore.setState("renderedItems", sortedItems);
    collection.setState("renderedItems", sortedItems);
  };
  _P5CWMRCEcjs.setup.call(void 0, collection, () => _P5CWMRCEcjs.init.call(void 0, privateStore));
  _P5CWMRCEcjs.setup.call(void 0, privateStore, () => {
    return _P5CWMRCEcjs.batch.call(void 0, privateStore, ["items"], (state) => {
      collection.setState("items", state.items);
    });
  });
  _P5CWMRCEcjs.setup.call(void 0, privateStore, () => {
    return _P5CWMRCEcjs.batch.call(void 0, privateStore, ["renderedItems"], (state) => {
      let firstRun = true;
      let raf = requestAnimationFrame(() => {
        const { renderedItems } = collection.getState();
        if (state.renderedItems === renderedItems) return;
        sortItems(state.renderedItems);
      });
      if (typeof IntersectionObserver !== "function") return () => cancelAnimationFrame(raf);
      const ioCallback = () => {
        if (firstRun) {
          firstRun = false;
          return;
        }
        cancelAnimationFrame(raf);
        raf = requestAnimationFrame(() => sortItems(state.renderedItems));
      };
      const root = getCommonParent(state.renderedItems);
      const observer = new IntersectionObserver(ioCallback, { root });
      for (const item of state.renderedItems) {
        if (!item.element) continue;
        observer.observe(item.element);
      }
      return () => {
        cancelAnimationFrame(raf);
        observer.disconnect();
      };
    });
  });
  const mergeItem = (item, { cache, setItems: setItems2, canDeleteFromMap = false }) => {
    let prevItem;
    setItems2((items2) => {
      const ids = !!cache.ids || items2.length >= itemIdCacheThreshold && (!canDeleteFromMap || !itemsMap.has(item.id)) ? getCachedItemIds(items2, cache) : void 0;
      const index = ids && !ids.has(item.id) ? -1 : items2.findIndex(({ id }) => id === item.id);
      const nextItems = items2.slice();
      if (index !== -1) {
        prevItem = items2[index];
        const nextItem = {
          ...prevItem,
          ...item
        };
        nextItems[index] = nextItem;
        itemsMap.set(item.id, nextItem);
      } else {
        nextItems.push(item);
        itemsMap.set(item.id, item);
        ids == null ? void 0 : ids.add(item.id);
      }
      if (cache.ids || index === -1 && nextItems.length >= itemIdCacheThreshold) setCachedItemIds(nextItems, cache, ids);
      return nextItems;
    });
    const unmergeItem = () => {
      setItems2((items2) => {
        const ids = cache.ids ? getCachedItemIds(items2, cache) : void 0;
        if (!prevItem) {
          if (canDeleteFromMap) itemsMap.delete(item.id);
          const nextItems2 = items2.filter(({ id }) => id !== item.id);
          ids == null ? void 0 : ids.delete(item.id);
          if (cache.ids) setCachedItemIds(nextItems2, cache, ids);
          return nextItems2;
        }
        const index = items2.findIndex(({ id }) => id === item.id);
        if (index === -1) return items2;
        const nextItems = items2.slice();
        nextItems[index] = prevItem;
        itemsMap.set(item.id, prevItem);
        if (cache.ids) setCachedItemIds(nextItems, cache, ids);
        return nextItems;
      });
    };
    return unmergeItem;
  };
  const setItems = (getItems) => privateStore.setState("items", getItems);
  const setRenderedItems = (getItems) => privateStore.setState("renderedItems", getItems);
  const itemMergeOptions = {
    cache: itemIdCache,
    setItems,
    canDeleteFromMap: true
  };
  const renderedItemMergeOptions = {
    cache: renderedItemIdCache,
    setItems: setRenderedItems
  };
  const registerItem = (item) => mergeItem(item, itemMergeOptions);
  return {
    ...collection,
    registerItem,
    renderItem: (item) => _GF35MDH7cjs.chain.call(void 0, registerItem(item), mergeItem(item, renderedItemMergeOptions)),
    item: (id) => {
      if (!id) return null;
      let item = itemsMap.get(id);
      if (!item) {
        const { items: items2 } = privateStore.getState();
        item = items2.find((item2) => item2.id === id);
        if (item) itemsMap.set(id, item);
      }
      return item || null;
    },
    __unstablePrivateStore: privateStore
  };
}

// ../ariakit-react-components/dist/collection/collection-store.js
function useCollectionStoreProps(store, update, props) {
  _GF35MDH7cjs.useUpdateEffect.call(void 0, update, [props.store]);
  _P5CWMRCEcjs.useStoreProps.call(void 0, store, props, "items", "setItems");
  return store;
}
function useCollectionStore(props = {}) {
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, createCollectionStore, props);
  return useCollectionStoreProps(store, update, props);
}









exports.useCollectionContext = useCollectionContext; exports.useCollectionProviderContext = useCollectionProviderContext; exports.CollectionContextProvider = CollectionContextProvider; exports.CollectionScopedContextProvider = CollectionScopedContextProvider; exports.createCollectionStore = createCollectionStore; exports.useCollectionStoreProps = useCollectionStoreProps; exports.useCollectionStore = useCollectionStore;
