// @ariakit/react/cjs/__chunks/LPRCUXZY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _BJQJHCU2cjs = require('./BJQJHCU2.cjs');















var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/button/utils.js
function withDefaultButtonType(props) {
  const { render, type } = props;
  if (render || type !== void 0) return props;
  return {
    ...props,
    type: "button"
  };
}

// ../ariakit-react-components/dist/command/command.js
var _react = require('react');
var TagName = "button";
function isNativeClick(event) {
  if (!event.isTrusted) return false;
  const element = event.currentTarget;
  if (event.key === "Enter") return _GF35MDH7cjs.isButton.call(void 0, element) || element.tagName === "SUMMARY" || element.tagName === "A";
  if (event.key === " ") return _GF35MDH7cjs.isButton.call(void 0, element) || element.tagName === "SUMMARY" || element.tagName === "INPUT" || element.tagName === "SELECT";
  return false;
}
var symbol = /* @__PURE__ */ Symbol("command");
var useCommand = _GF35MDH7cjs.createHook.call(void 0, function useCommand2({ clickOnEnter = true, clickOnSpace = true, ...props }) {
  const ref = _react.useRef.call(void 0, null);
  const [isNativeButton, setIsNativeButton] = _react.useState.call(void 0, false);
  const type = props.type;
  _react.useEffect.call(void 0, () => {
    const element = ref.current;
    if (!element) return;
    const nativeButton = _GF35MDH7cjs.isButton.call(void 0, element);
    if (type !== void 0 && nativeButton && element.type === "button") return;
    setIsNativeButton(nativeButton);
  }, [type]);
  const [active, setActive] = _react.useState.call(void 0, false);
  const activeRef = _react.useRef.call(void 0, false);
  const disabled = _GF35MDH7cjs.disabledFromProps.call(void 0, props);
  const [isDuplicate, metadataProps] = _GF35MDH7cjs.useMetadataProps.call(void 0, props, symbol, true);
  _GF35MDH7cjs.useSafeLayoutEffect.call(void 0, () => {
    if (!disabled) return;
    activeRef.current = false;
    setActive(false);
  }, [disabled]);
  const onKeyDownProp = props.onKeyDown;
  const onKeyDown = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    const element = event.currentTarget;
    if (event.defaultPrevented) return;
    if (isDuplicate) return;
    if (disabled) return;
    if (!_GF35MDH7cjs.isSelfTarget.call(void 0, event)) return;
    if (_GF35MDH7cjs.isTextField.call(void 0, element)) return;
    if (element.isContentEditable) return;
    const isEnter = clickOnEnter && event.key === "Enter";
    const isSpace = clickOnSpace && event.key === " ";
    const shouldPreventEnter = event.key === "Enter" && !clickOnEnter;
    const shouldPreventSpace = event.key === " " && !clickOnSpace;
    if (shouldPreventEnter || shouldPreventSpace) {
      event.preventDefault();
      return;
    }
    if (isEnter || isSpace) {
      const nativeClick = isNativeClick(event);
      if (isEnter) {
        if (!nativeClick) {
          event.preventDefault();
          const { view, ...eventInit } = event;
          const click = () => _GF35MDH7cjs.fireClickEvent.call(void 0, element, eventInit);
          if (_GF35MDH7cjs.isFirefox.call(void 0, )) _GF35MDH7cjs.queueBeforeEvent.call(void 0, element, "keyup", click);
          else queueMicrotask(click);
        }
      } else if (isSpace) {
        activeRef.current = true;
        if (!nativeClick) {
          event.preventDefault();
          setActive(true);
        }
      }
    }
  });
  const onKeyUpProp = props.onKeyUp;
  const onKeyUp = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onKeyUpProp == null ? void 0 : onKeyUpProp(event);
    if (isDuplicate) return;
    const isSpace = clickOnSpace && event.key === " ";
    if (!activeRef.current || !isSpace) return;
    const nativeClick = isNativeClick(event);
    activeRef.current = false;
    if (!nativeClick) setActive(false);
    if (event.defaultPrevented) return;
    if (!_GF35MDH7cjs.isSelfTarget.call(void 0, event)) return;
    if (disabled) return;
    if (event.metaKey) return;
    if (nativeClick) return;
    event.preventDefault();
    const element = event.currentTarget;
    const { view, ...eventInit } = event;
    queueMicrotask(() => _GF35MDH7cjs.fireClickEvent.call(void 0, element, eventInit));
  });
  const onBlurProp = props.onBlur;
  const onBlur = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onBlurProp == null ? void 0 : onBlurProp(event);
    if (!activeRef.current) return;
    activeRef.current = false;
    setActive(false);
  });
  props = {
    "data-active": active || void 0,
    type: isNativeButton ? "button" : void 0,
    ...metadataProps,
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref),
    onKeyDown,
    onKeyUp,
    onBlur
  };
  props = _BJQJHCU2cjs.useFocusable.call(void 0, props);
  return props;
});
var Command = _GF35MDH7cjs.forwardRef.call(void 0, function Command2(props) {
  const htmlProps = useCommand(withDefaultButtonType(props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});





exports.withDefaultButtonType = withDefaultButtonType; exports.useCommand = useCommand; exports.Command = Command;
