// @ariakit/react/cjs/__chunks/N6COOBTR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _DWSXHZYMcjs = require('./DWSXHZYM.cjs');






var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/heading/heading.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "h1";
var useHeading = _GF35MDH7cjs.createHook.call(void 0, function useHeading2(props) {
  const ref = _react.useRef.call(void 0, null);
  const level = _react.useContext.call(void 0, _DWSXHZYMcjs.HeadingContext) || 1;
  const Element = `h${level}`;
  const tagName = _GF35MDH7cjs.useTagName.call(void 0, ref, Element);
  const isNativeHeading = !!tagName && /^h\d$/.test(tagName);
  props = {
    render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Element, {}),
    role: !isNativeHeading ? "heading" : void 0,
    "aria-level": !isNativeHeading ? level : void 0,
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  return props;
});
var Heading = _GF35MDH7cjs.forwardRef.call(void 0, function Heading2(props) {
  const htmlProps = useHeading(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useHeading = useHeading; exports.Heading = Heading;
