// @ariakit/react/cjs/__chunks/S2WTSC3A.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _P52EZBZXcjs = require('./P52EZBZX.cjs');



var _4TCEVVD4cjs = require('./4TCEVVD4.cjs');
















var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/composite/composite-group.js
var TagName = "div";
var useCompositeGroup = _GF35MDH7cjs.createHook.call(void 0, function useCompositeGroup2({ store, ...props }) {
  props = _4TCEVVD4cjs.useGroup.call(void 0, props);
  return props;
});
var CompositeGroup = _GF35MDH7cjs.forwardRef.call(void 0, function CompositeGroup2(props) {
  const htmlProps = useCompositeGroup(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/composite/composite-group-label.js
var TagName2 = "div";
var useCompositeGroupLabel = _GF35MDH7cjs.createHook.call(void 0, function useCompositeGroupLabel2({ store, ...props }) {
  props = _4TCEVVD4cjs.useGroupLabel.call(void 0, props);
  return props;
});
var CompositeGroupLabel = _GF35MDH7cjs.forwardRef.call(void 0, function CompositeGroupLabel2(props) {
  const htmlProps = useCompositeGroupLabel(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});

// ../ariakit-react-components/dist/composite/composite-hover.js
var _react = require('react');
var TagName3 = "div";
function hoveringInside(event) {
  const nextElement = event.relatedTarget;
  if (!_GF35MDH7cjs.isElement.call(void 0, nextElement)) return false;
  return _GF35MDH7cjs.contains.call(void 0, event.currentTarget, nextElement);
}
var symbol = /* @__PURE__ */ Symbol("composite-hover");
function movingToAnotherItem(event) {
  const { relatedTarget } = event;
  if (!_GF35MDH7cjs.isElement.call(void 0, relatedTarget)) return false;
  let dest = relatedTarget;
  do {
    if (_GF35MDH7cjs.hasOwnProperty.call(void 0, dest, symbol) && dest[symbol]) return true;
    dest = dest.parentElement;
  } while (dest);
  return false;
}
var useCompositeHover = _GF35MDH7cjs.createHook.call(void 0, function useCompositeHover2({ store, focusOnHover = true, blurOnHoverEnd = !!focusOnHover, ...props }) {
  const context = _P52EZBZXcjs.useCompositeScopedContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "CompositeHover must be wrapped in a Composite component.");
  const isMouseMoving = _GF35MDH7cjs.useIsMouseMoving.call(void 0, );
  const onMouseMoveProp = props.onMouseMove;
  const focusOnHoverProp = _GF35MDH7cjs.useBooleanEvent.call(void 0, focusOnHover);
  const onMouseMove = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onMouseMoveProp == null ? void 0 : onMouseMoveProp(event);
    if (event.defaultPrevented) return;
    if (!isMouseMoving()) return;
    if (!focusOnHoverProp(event)) return;
    if (!_GF35MDH7cjs.hasFocusWithin.call(void 0, event.currentTarget)) {
      const baseElement = store == null ? void 0 : store.getState().baseElement;
      if (baseElement && !_GF35MDH7cjs.hasFocus.call(void 0, baseElement)) baseElement.focus();
    }
    store == null ? void 0 : store.setActiveId(event.currentTarget.id);
  });
  const onMouseLeaveProp = props.onMouseLeave;
  const blurOnHoverEndProp = _GF35MDH7cjs.useBooleanEvent.call(void 0, blurOnHoverEnd);
  const onMouseLeave = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    var _a;
    onMouseLeaveProp == null ? void 0 : onMouseLeaveProp(event);
    if (event.defaultPrevented) return;
    if (!isMouseMoving()) return;
    if (hoveringInside(event)) return;
    if (movingToAnotherItem(event)) return;
    if (!focusOnHoverProp(event)) return;
    if (!blurOnHoverEndProp(event)) return;
    store == null ? void 0 : store.setActiveId(null);
    (_a = store == null ? void 0 : store.getState().baseElement) == null ? void 0 : _a.focus();
  });
  const ref = _react.useCallback.call(void 0, (element) => {
    if (!element) return;
    element[symbol] = true;
  }, []);
  props = {
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref),
    onMouseMove,
    onMouseLeave
  };
  return _GF35MDH7cjs.removeUndefinedValues.call(void 0, props);
});
var CompositeHover = _GF35MDH7cjs.memo.call(void 0, _GF35MDH7cjs.forwardRef.call(void 0, function CompositeHover2(props) {
  const htmlProps = useCompositeHover(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName3, htmlProps);
}));








exports.useCompositeGroup = useCompositeGroup; exports.CompositeGroup = CompositeGroup; exports.useCompositeGroupLabel = useCompositeGroupLabel; exports.CompositeGroupLabel = CompositeGroupLabel; exports.useCompositeHover = useCompositeHover; exports.CompositeHover = CompositeHover;
