// @ariakit/react/cjs/__chunks/T2WGWXT5.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _KDGTJGN7cjs = require('./KDGTJGN7.cjs');



var _55MQXLFKcjs = require('./55MQXLFK.cjs');


var _SJZEGV7Acjs = require('./SJZEGV7A.cjs');






var _P52EZBZXcjs = require('./P52EZBZX.cjs');



var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');


















var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/toolbar/toolbar-context.js
var ctx = _GF35MDH7cjs.createStoreContext.call(void 0, [_P52EZBZXcjs.CompositeContextProvider], [_P52EZBZXcjs.CompositeScopedContextProvider]);
var useToolbarContext = ctx.useContext;
var useToolbarScopedContext = ctx.useScopedContext;
var useToolbarProviderContext = ctx.useProviderContext;
var ToolbarContextProvider = ctx.ContextProvider;
var ToolbarScopedContextProvider = ctx.ScopedContextProvider;

// ../ariakit-components/dist/toolbar/toolbar-store.js
function createToolbarStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  return _P52EZBZXcjs.createCompositeStore.call(void 0, {
    ...props,
    orientation: _GF35MDH7cjs.defaultValue.call(void 0, props.orientation, syncState == null ? void 0 : syncState.orientation, "horizontal"),
    focusLoop: _GF35MDH7cjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  });
}

// ../ariakit-react-components/dist/toolbar/toolbar-store.js
function useToolbarStoreProps(store, update, props) {
  return _55MQXLFKcjs.useCompositeStoreProps.call(void 0, store, update, props);
}
function useToolbarStore(props = {}) {
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, createToolbarStore, props);
  return useToolbarStoreProps(store, update, props);
}

// ../ariakit-react-components/dist/toolbar/toolbar.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useToolbar = _GF35MDH7cjs.createHook.call(void 0, function useToolbar2({ store: storeProp, orientation: orientationProp, virtualFocus, focusLoop, rtl, ...props }) {
  const context = useToolbarProviderContext();
  storeProp = storeProp || context;
  const store = useToolbarStore({
    store: storeProp,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl
  });
  const orientation = _P5CWMRCEcjs.useStoreState.call(void 0, store, ["orientation"], (state) => state.orientation === "both" ? void 0 : state.orientation);
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, ToolbarScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  props = {
    role: "toolbar",
    "aria-orientation": orientation,
    ...props
  };
  props = _55MQXLFKcjs.useComposite.call(void 0, {
    store,
    ...props
  });
  return props;
});
var Toolbar = _GF35MDH7cjs.forwardRef.call(void 0, function Toolbar2(props) {
  const htmlProps = useToolbar(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/composite/composite-container.js
var _react = require('react');
var TagName2 = "div";
function getFirstTabbable(container) {
  _GF35MDH7cjs.restoreFocusIn.call(void 0, container);
  const tabbable = _GF35MDH7cjs.getFirstTabbableIn.call(void 0, container);
  _GF35MDH7cjs.disableFocusIn.call(void 0, container);
  return tabbable;
}
var useCompositeContainer = _GF35MDH7cjs.createHook.call(void 0, function useCompositeContainer2({ store, ...props }) {
  const context = _P52EZBZXcjs.useCompositeScopedContext.call(void 0, );
  store = store || context;
  const ref = _react.useRef.call(void 0, null);
  const isOpenRef = _react.useRef.call(void 0, false);
  const open = (collapseToEnd = false) => {
    const container = ref.current;
    if (!container) return;
    _GF35MDH7cjs.restoreFocusIn.call(void 0, container);
    const tabbable = _GF35MDH7cjs.getFirstTabbableIn.call(void 0, container);
    if (!tabbable) {
      _GF35MDH7cjs.disableFocusIn.call(void 0, container);
      return;
    }
    isOpenRef.current = true;
    queueMicrotask(() => {
      tabbable.focus();
      if (_GF35MDH7cjs.isTextField.call(void 0, tabbable) || tabbable.isContentEditable) _P52EZBZXcjs.selectTextField.call(void 0, tabbable, collapseToEnd);
    });
  };
  const close = () => {
    const container = ref.current;
    if (!container) return;
    isOpenRef.current = false;
    _GF35MDH7cjs.disableFocusIn.call(void 0, container);
  };
  const renderedItems = _P5CWMRCEcjs.useStoreState.call(void 0, store, "renderedItems");
  _react.useEffect.call(void 0, () => {
    const container = ref.current;
    if (!container) return;
    if (!isOpenRef.current && (renderedItems == null ? void 0 : renderedItems.length)) _GF35MDH7cjs.disableFocusIn.call(void 0, container);
  }, [renderedItems]);
  const onFocusProp = props.onFocus;
  const onFocus = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onFocusProp == null ? void 0 : onFocusProp(event);
    if (event.defaultPrevented) return;
    if (!store) return;
    const isOpen = isOpenRef.current;
    if (_GF35MDH7cjs.isSelfTarget.call(void 0, event)) {
      isOpenRef.current = false;
      const { baseElement } = store.getState();
      const containers = baseElement == null ? void 0 : baseElement.querySelectorAll("[data-composite-container]");
      if (containers) for (const container of containers) _GF35MDH7cjs.disableFocusIn.call(void 0, container);
    } else if (!isOpen) {
      isOpenRef.current = true;
      _GF35MDH7cjs.restoreFocusIn.call(void 0, event.currentTarget);
      store == null ? void 0 : store.setState("moves", 0);
    }
  });
  const onBlurProp = props.onBlur;
  const onBlur = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onBlurProp == null ? void 0 : onBlurProp(event);
    if (event.defaultPrevented) return;
    if (_GF35MDH7cjs.isFocusEventOutside.call(void 0, event)) close();
  });
  const onKeyDownProp = props.onKeyDown;
  const onKeyDown = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.nativeEvent.isComposing) return;
    if (event.keyCode === 229) return;
    if (event.defaultPrevented) return;
    if (event.altKey) return;
    if (event.ctrlKey) return;
    if (event.metaKey) return;
    if (event.shiftKey) return;
    const container = event.currentTarget;
    if (_GF35MDH7cjs.isSelfTarget.call(void 0, event)) {
      if (event.key.length === 1 && event.key !== " ") {
        const tabbable = getFirstTabbable(container);
        if (!tabbable) return;
        if (_GF35MDH7cjs.isTextField.call(void 0, tabbable) || tabbable.isContentEditable) {
          event.stopPropagation();
          open();
        }
      } else if (event.key === "Delete" || event.key === "Backspace") {
        const tabbable = getFirstTabbable(container);
        if (!tabbable) return;
        if (_GF35MDH7cjs.isTextField.call(void 0, tabbable) || tabbable.isContentEditable) {
          open();
          const cleanup = () => {
            container.removeEventListener("input", onInput);
            container.removeEventListener("keyup", cleanup);
          };
          const onInput = () => {
            cleanup();
            queueMicrotask(() => container.focus());
          };
          container.addEventListener("input", onInput, { once: true });
          container.addEventListener("keyup", cleanup, { once: true });
          setTimeout(cleanup, 0);
        }
      }
    } else if (event.key === "Escape") queueMicrotask(() => container.focus());
    else if (event.key === "Enter") {
      const target = event.target;
      if (target.tagName === "INPUT" && !_GF35MDH7cjs.isButton.call(void 0, target) || target.tagName === "TEXTAREA" || target.isContentEditable) {
        event.preventDefault();
        queueMicrotask(() => container.focus());
      }
    }
  });
  const onClickProp = props.onClick;
  const onClick = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    if (_GF35MDH7cjs.isSelfTarget.call(void 0, event) && !event.detail) open(true);
  });
  props = {
    "data-composite-container": "",
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref),
    onFocus,
    onBlur,
    onKeyDown,
    onClick
  };
  return _GF35MDH7cjs.removeUndefinedValues.call(void 0, props);
});
var CompositeContainer = _GF35MDH7cjs.forwardRef.call(void 0, function CompositeContainer2(props) {
  const htmlProps = useCompositeContainer(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});

// ../ariakit-react-components/dist/toolbar/toolbar-item.js
var TagName3 = "button";
var useToolbarItem = _GF35MDH7cjs.createHook.call(void 0, function useToolbarItem2({ store, ...props }) {
  const context = useToolbarContext();
  store = store || context;
  props = _SJZEGV7Acjs.useCompositeItem.call(void 0, {
    store,
    ...props
  });
  return props;
});
var ToolbarItem = _GF35MDH7cjs.memo.call(void 0, _GF35MDH7cjs.forwardRef.call(void 0, function ToolbarItem2(props) {
  const htmlProps = useToolbarItem(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName3, htmlProps);
}));

// ../ariakit-react-components/dist/toolbar/toolbar-container.js
var TagName4 = "div";
var useToolbarContainer = _GF35MDH7cjs.createHook.call(void 0, function useToolbarContainer2({ store, ...props }) {
  const context = useToolbarContext();
  store = store || context;
  props = useCompositeContainer({
    store,
    ...props
  });
  props = useToolbarItem({
    store,
    ...props
  });
  return props;
});
var ToolbarContainer = _GF35MDH7cjs.memo.call(void 0, _GF35MDH7cjs.forwardRef.call(void 0, function ToolbarContainer2(props) {
  const htmlProps = useToolbarContainer(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName4, htmlProps);
}));

// ../ariakit-react-components/dist/toolbar/toolbar-input.js
var TagName5 = "input";
var useToolbarInput = _GF35MDH7cjs.createHook.call(void 0, function useToolbarInput2({ store, ...props }) {
  const context = useToolbarContext();
  store = store || context;
  props = useToolbarItem({
    store,
    ...props
  });
  return props;
});
var ToolbarInput = _GF35MDH7cjs.memo.call(void 0, _GF35MDH7cjs.forwardRef.call(void 0, function ToolbarInput2(props) {
  const htmlProps = useToolbarInput(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName5, htmlProps);
}));

// ../ariakit-react-components/dist/toolbar/toolbar-provider.js

function ToolbarProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, ToolbarContextProvider, {
    value: useToolbarStore(props),
    children: props.children
  });
}

// ../ariakit-react-components/dist/toolbar/toolbar-separator.js
var TagName6 = "hr";
var useToolbarSeparator = _GF35MDH7cjs.createHook.call(void 0, function useToolbarSeparator2({ store, ...props }) {
  const context = useToolbarContext();
  store = store || context;
  props = _KDGTJGN7cjs.useCompositeSeparator.call(void 0, {
    store,
    ...props
  });
  return props;
});
var ToolbarSeparator = _GF35MDH7cjs.forwardRef.call(void 0, function ToolbarSeparator2(props) {
  const htmlProps = useToolbarSeparator(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName6, htmlProps);
});










exports.useToolbarContext = useToolbarContext; exports.useToolbarStore = useToolbarStore; exports.Toolbar = Toolbar; exports.ToolbarItem = ToolbarItem; exports.ToolbarContainer = ToolbarContainer; exports.ToolbarInput = ToolbarInput; exports.ToolbarProvider = ToolbarProvider; exports.ToolbarSeparator = ToolbarSeparator;
