// @ariakit/react/cjs/__chunks/TB5IKUMV.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _KAU3JWAXcjs = require('./KAU3JWAX.cjs');



var _3PZT7FSKcjs = require('./3PZT7FSK.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');




var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/popover/popover-dismiss.js
var TagName = "button";
var usePopoverDismiss = _GF35MDH7cjs.createHook.call(void 0, function usePopoverDismiss2({ store, ...props }) {
  const context = _KAU3JWAXcjs.usePopoverScopedContext.call(void 0, );
  store = store || context;
  props = _3PZT7FSKcjs.useDialogDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var PopoverDismiss = _GF35MDH7cjs.forwardRef.call(void 0, function PopoverDismiss2(props) {
  const htmlProps = usePopoverDismiss(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/popover/popover-heading.js
var TagName2 = "h1";
var usePopoverHeading = _GF35MDH7cjs.createHook.call(void 0, function usePopoverHeading2(props) {
  props = _3PZT7FSKcjs.useDialogHeading.call(void 0, props);
  return props;
});
var PopoverHeading = _GF35MDH7cjs.forwardRef.call(void 0, function PopoverHeading2(props) {
  const htmlProps = usePopoverHeading(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});






exports.usePopoverDismiss = usePopoverDismiss; exports.PopoverDismiss = PopoverDismiss; exports.usePopoverHeading = usePopoverHeading; exports.PopoverHeading = PopoverHeading;
