// @ariakit/react/cjs/__chunks/VMD6DM5U.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _EHTBB6EYcjs = require('./EHTBB6EY.cjs');


var _F7MQUJ3Qcjs = require('./F7MQUJ3Q.cjs');



var _TB5IKUMVcjs = require('./TB5IKUMV.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');




var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/hovercard/hovercard-description.js
var TagName = "p";
var useHovercardDescription = _GF35MDH7cjs.createHook.call(void 0, function useHovercardDescription2(props) {
  props = _F7MQUJ3Qcjs.usePopoverDescription.call(void 0, props);
  return props;
});
var HovercardDescription = _GF35MDH7cjs.forwardRef.call(void 0, function HovercardDescription2(props) {
  const htmlProps = useHovercardDescription(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/hovercard/hovercard-dismiss.js
var TagName2 = "button";
var useHovercardDismiss = _GF35MDH7cjs.createHook.call(void 0, function useHovercardDismiss2({ store, ...props }) {
  const context = _EHTBB6EYcjs.useHovercardScopedContext.call(void 0, );
  store = store || context;
  props = _TB5IKUMVcjs.usePopoverDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var HovercardDismiss = _GF35MDH7cjs.forwardRef.call(void 0, function HovercardDismiss2(props) {
  const htmlProps = useHovercardDismiss(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});

// ../ariakit-react-components/dist/hovercard/hovercard-heading.js
var TagName3 = "h1";
var useHovercardHeading = _GF35MDH7cjs.createHook.call(void 0, function useHovercardHeading2(props) {
  props = _TB5IKUMVcjs.usePopoverHeading.call(void 0, props);
  return props;
});
var HovercardHeading = _GF35MDH7cjs.forwardRef.call(void 0, function HovercardHeading2(props) {
  const htmlProps = useHovercardHeading(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName3, htmlProps);
});








exports.useHovercardDescription = useHovercardDescription; exports.HovercardDescription = HovercardDescription; exports.useHovercardDismiss = useHovercardDismiss; exports.HovercardDismiss = HovercardDismiss; exports.useHovercardHeading = useHovercardHeading; exports.HovercardHeading = HovercardHeading;
