// @ariakit/react/cjs/__chunks/YHJUBL36.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _55MQXLFKcjs = require('./55MQXLFK.cjs');




var _P52EZBZXcjs = require('./P52EZBZX.cjs');



var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');







var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/menubar/menubar-context.js
var menubar = _GF35MDH7cjs.createStoreContext.call(void 0, [_P52EZBZXcjs.CompositeContextProvider], [_P52EZBZXcjs.CompositeScopedContextProvider]);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;

// ../ariakit-components/dist/menubar/menubar-store.js
function createMenubarStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  return _P52EZBZXcjs.createCompositeStore.call(void 0, {
    ...props,
    orientation: _GF35MDH7cjs.defaultValue.call(void 0, props.orientation, syncState == null ? void 0 : syncState.orientation, "horizontal"),
    focusLoop: _GF35MDH7cjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true)
  });
}

// ../ariakit-react-components/dist/menubar/menubar-store.js
function useMenubarStoreProps(store, update, props) {
  return _55MQXLFKcjs.useCompositeStoreProps.call(void 0, store, update, props);
}
function useMenubarStore(props = {}) {
  const [store, update] = _P5CWMRCEcjs.useStore.call(void 0, createMenubarStore, props);
  return useMenubarStoreProps(store, update, props);
}

// ../ariakit-react-components/dist/menubar/menubar.js
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useMenubar = _GF35MDH7cjs.createHook.call(void 0, function useMenubar2({ store: storeProp, composite = true, orientation: orientationProp, virtualFocus, focusLoop, rtl, ...props }) {
  const context = useMenubarProviderContext();
  storeProp = storeProp || context;
  const store = useMenubarStore({
    store: storeProp,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl
  });
  const orientation = _P5CWMRCEcjs.useStoreState.call(void 0, store, ["orientation"], (state) => !composite || state.orientation === "both" ? void 0 : state.orientation);
  props = _GF35MDH7cjs.useWrapElement.call(void 0, props, (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, MenubarScopedContextProvider, {
    value: store,
    children: element
  }), [store]);
  if (composite) props = {
    role: "menubar",
    "aria-orientation": orientation,
    ...props
  };
  props = _55MQXLFKcjs.useComposite.call(void 0, {
    store,
    composite,
    ...props
  });
  return props;
});
var Menubar = _GF35MDH7cjs.forwardRef.call(void 0, function Menubar2(props) {
  const htmlProps = useMenubar(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/menubar/menubar-provider.js

function MenubarProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, MenubarContextProvider, {
    value: useMenubarStore(props),
    children: props.children
  });
}








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarStore = useMenubarStore; exports.useMenubar = useMenubar; exports.Menubar = Menubar; exports.MenubarProvider = MenubarProvider;
