// @ariakit/react/cjs/__chunks/YWXNAW4U.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});




var _EHTBB6EYcjs = require('./EHTBB6EY.cjs');


var _J27UXORVcjs = require('./J27UXORV.cjs');


var _NBWQLKS5cjs = require('./NBWQLKS5.cjs');


var _ERZLFUVIcjs = require('./ERZLFUVI.cjs');


var _P5CWMRCEcjs = require('./P5CWMRCE.cjs');


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');











var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/hovercard/hovercard-arrow.js
var TagName = "div";
var useHovercardArrow = _GF35MDH7cjs.createHook.call(void 0, function useHovercardArrow2({ store, ...props }) {
  const context = _EHTBB6EYcjs.useHovercardContext.call(void 0, );
  store = store || context;
  props = _J27UXORVcjs.usePopoverArrow.call(void 0, {
    store,
    ...props
  });
  return props;
});
var HovercardArrow = _GF35MDH7cjs.forwardRef.call(void 0, function HovercardArrow2(props) {
  const htmlProps = useHovercardArrow(props);
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});

// ../ariakit-react-components/dist/hovercard/hovercard-disclosure.js
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName2 = "button";
var useHovercardDisclosure = _GF35MDH7cjs.createHook.call(void 0, function useHovercardDisclosure2({ store, ...props }) {
  const context = _EHTBB6EYcjs.useHovercardProviderContext.call(void 0, );
  store = store || context;
  _GF35MDH7cjs.invariant.call(void 0, store, process.env.NODE_ENV !== "production" && "HovercardDisclosure must receive a `store` prop or be wrapped in a HovercardProvider component.");
  const [visible, setVisible] = _react.useState.call(void 0, false);
  _react.useEffect.call(void 0, () => {
    if (!visible) return;
    const onBlur = (event) => {
      if (!store) return;
      const nextActiveElement = event.relatedTarget;
      if (_GF35MDH7cjs.isNode.call(void 0, nextActiveElement)) {
        const { anchorElement: anchor, popoverElement: popover, disclosureElement: disclosure } = store.getState();
        if (anchor && _GF35MDH7cjs.contains.call(void 0, anchor, nextActiveElement)) return;
        if (popover && _GF35MDH7cjs.contains.call(void 0, popover, nextActiveElement)) return;
        if (disclosure && _GF35MDH7cjs.contains.call(void 0, disclosure, nextActiveElement)) return;
        if (_GF35MDH7cjs.isElement.call(void 0, nextActiveElement) && nextActiveElement.hasAttribute("data-focus-trap")) return;
      }
      setVisible(false);
    };
    return _GF35MDH7cjs.addGlobalEventListener.call(void 0, "focusout", onBlur, true);
  }, [visible, store]);
  _react.useEffect.call(void 0, () => {
    return _P5CWMRCEcjs.sync.call(void 0, store, ["anchorElement"], (state) => {
      const anchor = state.anchorElement;
      if (!anchor) return;
      const observer = new MutationObserver(() => {
        if (!anchor.hasAttribute("data-focus-visible")) return;
        setVisible(true);
      });
      observer.observe(anchor, { attributeFilter: ["data-focus-visible"] });
      return () => observer.disconnect();
    });
  }, [store]);
  const onClickProp = props.onClick;
  const onClick = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    store == null ? void 0 : store.setAutoFocusOnShow(true);
  });
  const onFocusProp = props.onFocus;
  const onFocus = _GF35MDH7cjs.useEvent.call(void 0, (event) => {
    onFocusProp == null ? void 0 : onFocusProp(event);
    if (event.defaultPrevented) return;
    setVisible(true);
  });
  const { style } = _ERZLFUVIcjs.useVisuallyHidden.call(void 0, );
  if (!visible) props = {
    ...props,
    style: {
      ...style,
      ...props.style
    }
  };
  props = {
    children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "svg", {
      display: "block",
      fill: "none",
      stroke: "currentColor",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.5,
      viewBox: "0 0 16 16",
      height: "1em",
      width: "1em",
      children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points: "4,6 8,10 12,6" })
    }),
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, store.setDisclosureElement, props.ref),
    onClick,
    onFocus
  };
  props = _NBWQLKS5cjs.useDialogDisclosure.call(void 0, {
    store,
    ...props
  });
  return props;
});
var HovercardDisclosure = _GF35MDH7cjs.forwardRef.call(void 0, function HovercardDisclosure2(props) {
  const htmlProps = useHovercardDisclosure(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName2, htmlProps);
});

// ../ariakit-react-components/dist/hovercard/hovercard-provider.js

function HovercardProvider(props = {}) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _EHTBB6EYcjs.HovercardContextProvider, {
    value: _EHTBB6EYcjs.useHovercardStore.call(void 0, props),
    children: props.children
  });
}





exports.HovercardArrow = HovercardArrow; exports.HovercardDisclosure = HovercardDisclosure; exports.HovercardProvider = HovercardProvider;
