// @ariakit/react/cjs/__chunks/ZBISR5IR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _KAU3JWAXcjs = require('./KAU3JWAX.cjs');



var _P52EZBZXcjs = require('./P52EZBZX.cjs');


var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/combobox/combobox-context.js
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, void 0);
var ctx = _GF35MDH7cjs.createStoreContext.call(void 0, [_KAU3JWAXcjs.PopoverContextProvider, _P52EZBZXcjs.CompositeContextProvider], [_KAU3JWAXcjs.PopoverScopedContextProvider, _P52EZBZXcjs.CompositeScopedContextProvider]);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, void 0);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
