// @ariakit/react/cjs/__chunks/ZQUU53RK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _LPRCUXZYcjs = require('./LPRCUXZY.cjs');







var _GF35MDH7cjs = require('./GF35MDH7.cjs');

// ../ariakit-react-components/dist/button/button.js
var _react = require('react');
var TagName = "button";
var useButton = _GF35MDH7cjs.createHook.call(void 0, function useButton2(props) {
  const ref = _react.useRef.call(void 0, null);
  const tagName = _GF35MDH7cjs.useTagName.call(void 0, ref, TagName);
  const [isNativeButton, setIsNativeButton] = _react.useState.call(void 0, () => !!tagName && _GF35MDH7cjs.isButton.call(void 0, {
    tagName,
    type: props.type
  }));
  _react.useEffect.call(void 0, () => {
    if (!ref.current) return;
    setIsNativeButton(_GF35MDH7cjs.isButton.call(void 0, ref.current));
  }, []);
  props = {
    role: !isNativeButton && tagName !== "a" ? "button" : void 0,
    ...props,
    ref: _GF35MDH7cjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  props = _LPRCUXZYcjs.useCommand.call(void 0, props);
  return props;
});
var Button = _GF35MDH7cjs.forwardRef.call(void 0, function Button2(props) {
  const htmlProps = useButton(_LPRCUXZYcjs.withDefaultButtonType.call(void 0, props));
  return _GF35MDH7cjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useButton = useButton; exports.Button = Button;
